mesonbuild/modules/i18n.py

```diff
@@ -14,47 +14,245 @@
+                             'merge_files': self.merge_files_method,
+                             'add_merge_files': self.add_merge_files_method,
+                             })
+
+    def add_files_method(self, args, kwargs):
+        self.held_object.add_files(args, kwargs)
+
+    def merge_files_method(self, args, kwargs):
+        targets = self.held_object.merge_files(args, kwargs)
+        self.held_object.interpreter.process_new_values(targets)
+        return targets
+
+    def add_merge_files_method(self, args, kwargs):
+        if len(args) < 1:
+            raise coredata.MesonException('add_merge_files requires at least one positional argument (file).')
+        if 'extra_args' in kwargs:
+            raise coredata.MesonException('add_merge_files does not accept extra_args.')
+        self.held_object.add_files(args, kwargs.copy())
+        targets = self.held_object.merge_files(args, kwargs)
+        self.held_object.interpreter.process_new_values(targets)
+        return targets
+
+
+class I18nPotGenerator:
+
+    @staticmethod
+    def _get_keywords(kwargs, preset):
+        keywords = mesonlib.stringlistify(kwargs.pop('keywords', []))
+        for k in preset['keywords']:
+            if k not in keywords:
+                keywords.append(k)
+        return keywords
+
+    def _get_preset(self, kwargs):
+        preset = kwargs.pop('preset', self.preset)
+        if preset not in PRESET_ARGS:
+            raise coredata.MesonException('i18n: Preset "{}" is not one of the valid options: {}'.format(
+                                          preset, list(PRESET_ARGS.keys())))
+        return PRESET_ARGS[preset]
+
+    def __init__(self, interpreter, args, kwargs):
+        super().__init__()
+
+        if len(args) != 1:
+            raise coredata.MesonException('Gettext requires one positional argument (package name).')
+        if not shutil.which('xgettext'):
+            raise coredata.MesonException('Can not do gettext because xgettext is not installed.')
+
+        self.interpreter = interpreter
+        self.environment = interpreter.environment
+        self.subdir = interpreter.subdir
+        packagename = args[0]
+        languages = mesonlib.stringlistify(kwargs.get('languages', []))
+        self.datadirs = I18nModule._get_data_dirs(self.environment, self.subdir, mesonlib.stringlistify(kwargs.get('data_dirs', [])))
+
+        self.preset = kwargs.pop('preset', 'none')
+        if self.preset not in PRESET_ARGS:
+            raise coredata.MesonException('i18n: Preset "{}" is not one of the valid options: {}'.format(
+                                          self.preset, list(PRESET_ARGS.keys())))
+
+        pkg_arg = '--pkgname=' + packagename
+        lang_arg = '--langs=' + '@@'.join(languages) if languages else None
+        datadirs = '--datadirs=' + ':'.join(self.datadirs) if self.datadirs else None
+        pot_data_file = None
+        extra_args = None
+        preset = self._get_preset(kwargs)
```

Comment:
__[flake8]__
* [F841] local variable 'preset' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->